net: wifi_mgmt: add hostap DPP support

fengming-ye opened the pull request:
Add wifi l2 mgmt dpp handlers.
Add wifi subcommand dpp to call l2 mgmt dpp handlers.

DPP shell commands body is the same as wpa_cli commands but with different prefix "wifi dpp". eg.
"wifi dpp dpp_auth_init peer=1 role=enrollee"

DPP shell command handler will parse user args to params in enum and send l2 mgmt dpp requests.
DPP l2 handlers will parse params to hostap wpa_cli format args and send wpa_cli commands to hostap.

fengming-ye committed:
net: wifi_mgmt: add hostap DPP support

Add wifi l2 mgmt dpp handlers.
Add wifi subcommand dpp to call l2 mgmt dpp handlers.

DPP l2 handlers will parse params to hostap wpa_cli format args
and send wpa_cli commands to hostap.

Signed-off-by: Fengming Ye <frank.ye@nxp.com>

fengming-ye commented:
Complicance check fail because of unknown kconfig CONFIG_WIFI_NM_WPA_SUPPLICANT_DPP
Because it relies on following PR
https://github.com/zephyrproject-rtos/zephyr/pull/73249

Pls help review regardless of this CI error so that this PR can go in parallel.

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +777,105 @@ struct wifi_ap_config_params {
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_DPP
+#define WIFI_DPP_QRCODE_MAX_LEN 255
+
+enum wifi_dpp_op {
```
You need to document these using doxygen comments `/** .... */`, same comment to other added symbols.

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +777,105 @@ struct wifi_ap_config_params {
 };
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_DPP
+#define WIFI_DPP_QRCODE_MAX_LEN 255
```
Also this define needs documentation

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +777,105 @@ struct wifi_ap_config_params {
 	uint32_t max_num_sta;
 };
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_DPP
```
No need to hide these enums, structs and defines behind the Kconfig option, it is much easier to work with the code if these can be freely used.

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -1008,3 +1008,269 @@ int supplicant_ap_sta_disconnect(const struct device *dev,
 }
 #endif /* CONFIG_AP */
+
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_DPP
```
You can remove this so that the DPP code gets compile tested even if it is not used.

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -1008,3 +1008,269 @@ int supplicant_ap_sta_disconnect(const struct device *dev,
 #endif /* CONFIG_AP */
+
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_DPP
+static char *dpp_params_to_args_curve(int curve)
```
Please return `static const char *...` so that the strings get placed to ROM.

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -1008,3 +1008,269 @@ int supplicant_ap_sta_disconnect(const struct device *dev,
+	}
+}
+
+static char *dpp_params_to_args_conf(int conf)
```
same here

fengming-ye commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1933,6 +1933,312 @@ static int cmd_wifi_version(const struct shell *sh, size_t argc, char *argv[])
+
+		pos = strstr(argv[i], "chan=");
+		if (pos) {
+			params->bootstrap_gen.op_class = atoi(pos + strlen("chan="));
```
Updated as comments

jukkar commented:
This PR seems to be related to this hostap PR https://github.com/zephyrproject-rtos/hostap/pull/10
It is important that we link these PRs together so that things get at least compile tested before we merge the PRs.
So in this PR, you should add a commit that adds the hostap PR to west.yml manifest file.
The hostap revision should be set to `revision: pull/10/head` and then you should re-submit this PR. After the hostap PR is merged, you need to send one more push of this PR and change the revision to the actual head commit ID of the hostap repo.
This will mean one extra round of tests but that is unavoidable here.

jukkar commented:
The hostap PRs are now merged, next steps:
* update west.yml hostap revision to commit 20f4cacf5a69eed9bc684f11c62b8f75349f223a which is latest in upstream
* it might make sense to wait until the upstream has fixed the sensing subsystem issue which is causing CI issues before basing and re-submitting the pr, the hotfix should come still today

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +776,124 @@ struct wifi_ap_config_params {
+
+/** Wi-Fi DPP operations */
+enum wifi_dpp_op {
+	WIFI_DPP_OP_INVALID = 0,
+	WIFI_DPP_CONFIGURATOR_ADD,
+	WIFI_DPP_AUTH_INIT,
+	WIFI_DPP_QR_CODE,
+	WIFI_DPP_CHIRP,
+	WIFI_DPP_LISTEN,
+	WIFI_DPP_BOOTSTRAP_GEN,
+	WIFI_DPP_BOOTSTRAP_GET_URI,
+	WIFI_DPP_SET_CONF_PARAM,
+	WIFI_DPP_SET_WAIT_RESP_TIME
```
Please document these operations. If the operations are more or less internal and not exposed to users or applications, then we could also mark them as internal with
```
/** @cond INTERNAL_HIDDEN */
...
/** @endcond */
```
which prevents the comments from generated documentation. See examples how this is used in this file for example.

fengming-ye replied:
These enums are external. But I think these enum names already explains themselves.
I can add comments for them. But these comments will be similar to names. It that okay?

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +776,124 @@ struct wifi_ap_config_params {
+
+/** Wi-Fi DPP crypto curves, default using prime256v1 */
+enum wifi_dpp_curves {
+	WIFI_DPP_CURVES_DEFAULT = 0,
+	WIFI_DPP_CURVES_P_256,
+	WIFI_DPP_CURVES_P_384,
+	WIFI_DPP_CURVES_P_512,
+	WIFI_DPP_CURVES_BP_256,
+	WIFI_DPP_CURVES_BP_384,
+	WIFI_DPP_CURVES_BP_512
```
Same here

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +776,124 @@ struct wifi_ap_config_params {
+ * either allows both roles
+ */
+enum wifi_dpp_role {
+	WIFI_DPP_ROLE_UNSET = 0,
+	WIFI_DPP_ROLE_CONFIGURATOR,
+	WIFI_DPP_ROLE_ENROLLEE,
+	WIFI_DPP_ROLE_EITHER
```
And here

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +776,124 @@ struct wifi_ap_config_params {
+ * current only support DPP only AKM
+ */
+enum wifi_dpp_conf {
+	WIFI_DPP_CONF_UNSET = 0,
+	WIFI_DPP_CONF_STA,
+	WIFI_DPP_CONF_AP,
+	WIFI_DPP_CONF_QUERY
```
Here too

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +776,124 @@ struct wifi_ap_config_params {
+ * current default and only support QR-CODE
+ */
+enum wifi_dpp_bootstrap_type {
+	WIFI_DPP_BOOTSTRAP_TYPE_UNSET = 0,
+	WIFI_DPP_BOOTSTRAP_TYPE_QRCODE,
+	WIFI_DPP_BOOTSTRAP_TYPE_PKEX,
+	WIFI_DPP_BOOTSTRAP_TYPE_NFC_URI
```
And here

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +776,124 @@ struct wifi_ap_config_params {
+			int curve;
+			int net_access_key_curve;
+		} configurator_add;
+		struct wifi_dpp_auth_init_params {
+			int peer;
+			int configurator;
+			int role;
+			int conf;
+			char ssid[WIFI_SSID_MAX_LEN + 1];
+		} auth_init;
+		struct wifi_dpp_chirp_params {
+			int id;
+			int freq;
+		} chirp;
+		struct wifi_dpp_listen_params {
+			int freq;
+			int role;
+		} listen;
+		struct wifi_dpp_bootstrap_gen_params {
+			int op_class;
+			int chan;
+			int curve;
+			int type;
+			uint8_t mac[WIFI_MAC_ADDR_LEN];
+		} bootstrap_gen;
+		struct wifi_dpp_configurator_set_params {
+			int peer;
+			int configurator;
+			int role;
+			int conf;
+			int curve;
+			int net_access_key_curve;
+			char ssid[WIFI_SSID_MAX_LEN + 1];
+		} configurator_set;
+		/* bootstrap get uri id */
+		int id;
+		/* timeout for dpp frame response rx */
+		int dpp_resp_wait_time;
+		/* dpp QR-CODE, max for SHA512 */
+		uint8_t dpp_qr_code[WIFI_DPP_QRCODE_MAX_LEN + 1];
```
All of these fields need proper doxygen comments

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -1008,3 +1008,267 @@ int supplicant_ap_sta_disconnect(const struct device *dev,
+
+	switch (params->action) {
+	case WIFI_DPP_CONFIGURATOR_ADD:
+		strcpy(pos, "dpp_configurator_add");
```
You should use `strncpy()` variant when copying. This way we can be sure that we will not overwrite `dpp_cmd_buf. There are external tools that will complain the use of these unsafe functions so we could avoid future issues for these.

fengming-ye replied:
updated as comments

jukkar commented on west.yml:
```diff
@@ -261,7 +261,7 @@ manifest:
       repo-path: hostap
       path: modules/lib/hostap
-      revision: 83574f533fb5c6808af0d9872741d72d48be2f98
+      revision: pull/10/head
```
This can be changed to 20f4cacf5a69eed9bc684f11c62b8f75349f223a as that is now the latest version of hostap repo.

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -1008,3 +1008,268 @@ int supplicant_ap_sta_disconnect(const struct device *dev,
+
+	switch (params->action) {
+	case WIFI_DPP_CONFIGURATOR_ADD:
+		strncpy(pos, "dpp_configurator_add", sizeof("dpp_configurator_add") - 1);
```
The 3rd parameter should be the remaining length of the pos so that we do not overwrite it

fengming-ye replied:
updated along with new commit to add response parameter in dpp params, to get and parse resp for L2 caller function.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2122,6 +2479,24 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		     0),
+	SHELL_CMD_ARG(dpp,
+		      NULL,
+		      "DPP actions refer to hostap\n"
+		      "dpp_configurator_add [curve=P-256] [net_access_key_curve=P-256]\n"
+		      "dpp_auth_init [peer=<peer_bi_id>] [conf=sta-dpp/ap-dpp]"
+		      " [ssid=<SSID>] [configurator=<configurator_id>]"
+		      " [role=<configurator/enrollee/either>]"
+		      "dpp_qr_code <qr_code>\n"
+		      "dpp_chirp own=<own_bi_id> listen=<freq>\n"
+		      "dpp_listen <freq> role=<configurator/enrollee/either>\n"
+		      "dpp_bootstrap_gen type=qrcode chan=<opclass/channel> mac=<mac addr>"
+		      " [curve=P-256]\n"
+		      "set dpp_configurator_params [peer=peer_bi_id] [conf=sta-dpp] [ssid=<SSID>]"
+		      " [configurator=1] [curve=P-256] [net_access_key_curve=P-256]\n"
+		      "set dpp_resp_wait_time <time_ms>\n",
```
for new command we shifted to usign `getopt` as it makes parsing and entering commands easier.

fengming-ye replied:
DPP commands are simply tight to wpa_cli format.
That's easier for user to switch from linux wpa_supplicant to zephyr wpa_supplicant.

fengming-ye replied:
so is this okay this way for DPP?

krish2718 replied:
Sure, but in general we are not maintaining compatibility of wpa_cli vs wifi shell, for non-user cases like Certifications you can use `wpa_cli` directly https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hostap/src/wpa_cli.c#L17 and native Zephyr commands are aligned with Zephyr usage.

E.g., a typical DPP usage in Zephyr may not need all these low-level commands, so, you can add an abstraction to limit/document these, so, `wifi shell` is more controlled and `wpa_cli` takes any supported command.

fengming-ye replied:
yes this dose ring a bell. I can restructure dpp shell to a more reasonable way by args with '-' and in subcommands,
which may be easier to extend.
I'll change this commit ASAP.

fengming-ye replied:
updated wifi_shell.c as this comment requierd

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +776,172 @@ struct wifi_ap_config_params {
+		} configurator_set;
+		/** Bootstrap get uri id */
+		int id;
+		/** Timeout for dpp frame response rx */
```
DPP in comments.

fengming-ye replied:
updated

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -1091,7 +1091,7 @@ int supplicant_dpp_dispatch(const struct device *dev,
 	switch (params->action) {
 	case WIFI_DPP_CONFIGURATOR_ADD:
-		strncpy(pos, "dpp_configurator_add", end - pos);
+		strncpy(pos, "DPP_CONFIGURATOR_ADD", end - pos);
```
fixup commits aren't allowed, they should be fixed up with the original commit.

fengming-ye replied:
added in original commit

krish2718 commented on west.yml:
```diff
@@ -261,7 +261,7 @@ manifest:
       repo-path: hostap
       path: modules/lib/hostap
-      revision: 83574f533fb5c6808af0d9872741d72d48be2f98
+      revision: pull/12/head
```
This should be the `oldest` commit as without this other commits won't build.

fengming-ye replied:
not quite understand this comment.
https://github.com/zephyrproject-rtos/hostap/pull/12 is new PR on top of latest commit, I think.

krish2718 replied:
the ordering would be (old to new):

* Bring in changes from hostap as these are needed before we use them e.g., a new API added
* Do the corresponding changes in `modules/hostap` e.g., call new API

this way build isn't broken if someone checks out either commit. But if the changes have circular dependency, this cannot be fixed and order doesn't matter then.

jukkar commented:
@dleach02 do you need this to 3.7, if yes would you present this to TSC?

dleach02 commented:
> @dleach02 do you need this to 3.7, if yes would you present this to TSC?

While it would be nice to include, there are others that would be needed as well so we are planning on supporting on our own downstream.

krish2718 commented:
@fengming-ye please resolve conflicts and also rebase.

fabiobaltieri commented:
merge conflict, can you rebase please?

aescolar commented:
@fengming-ye Rebase needed

fengming-ye commented:
> @fengming-ye Rebase needed

https://github.com/zephyrproject-rtos/zephyr/pull/76454
rebased on this PR

jukkar commented:
> #76454
> rebased on this PR

This is very confusing if you create a new PR, please just use this one and just force push new commit versions on top of this PR.

jukkar commented:
For this PR, just rebase your branch to latest main, and force push. Github complains now about some conflict which hopefully is resolved by a rebase.

fengming-ye commented:
CI fail because of unrelated error

INFO    - 659/659 qemu_x86                  tests/net/lib/http_server/tls/net.http.server.tls   FAILED unexpected eof (qemu 0.551s)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_x86/tests/net/lib/http_server/tls/net.http.server.tls/handler.log
ERROR   - SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build v3.7.0-310-gc326e014cab1 ***
W: No entropy device on the system, TLS communication is insecure!
Running TESTSUITE framework_tests_tls
===================================================================
START - test_tls
E: Page fault at address 0x18a00c (error code 0x0)
E: Linear address not present in page tables
E: Access violation: supervisor thread not allowed to read
E: PTE: not present
E: EAX: 0x00139180, EBX: 0x0018a008, ECX: 0x00000059, EDX: 0x00000000
E: ESI: 0x00000000, EDI: 0x0018a010, EBP: 0x007e5e8c, ESP: 0x007e5e78
E: EFLAGS: 0x00000283 CS: 0x0008 CR3: 0x0014a000
E: call trace:
E: EIP: 0x0011e171
E:      0x0011e974 (0x14f1f4)
E:      0x0011df4a (0x14eed0)
E:      0x0010134c (0x0)
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x14ee00 (http_server_tid)
E: Halting system

jukkar commented:
The CI issue is already fixed by #76580. Please rebase against latest main and force push a new version.

fengming-ye commented:
> The CI issue is already fixed by #76580. Please rebase against latest main and force push a new version.

rebased and add NXP copyright since we are adding new features to wifi mgmt.

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +777,179 @@ struct wifi_ap_config_params {
+
+/** Wi-Fi DPP params for various operations
+ *
+ * peer peer bootstrap id
+ * id own bootstrap id
+ * configurator configurator id
+ * ssid ssid in string
```
What are these comments referring to? I mean the individual fields are already commented below so these four lines seems unnecessary.

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +777,179 @@ struct wifi_ap_config_params {
+ * ssid ssid in string
+ */
+struct wifi_dpp_params {
+	/** operation enum */
```
Please use proper English here as this data is used to generate the documentation, so start with capital letter here "Operation enum"

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +777,179 @@ struct wifi_ap_config_params {
+	/** operation enum */
+	int action;
+	union {
+		struct wifi_dpp_configurator_add_params {
```
The structs should also have a doxygen comment.
```
/** xyz.... */
struct wifi_dpp_..... {
```
and so on

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -768,6 +777,179 @@ struct wifi_ap_config_params {
+		int dpp_resp_wait_time;
+		/** DPP QR-CODE, max for SHA512 */
+		uint8_t dpp_qr_code[WIFI_DPP_QRCODE_MAX_LEN + 1];
+		/** request response
+		 *
+		 * in union to save runtime stack,
+		 * when resp is fulfilled, request params should
+		 * no longer be used
```
This looks a bit weird, could you rewrite it.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1,5 +1,6 @@
 /*
```
Could you place the shell changes to a separate commit

fengming-ye replied:
updated as above requests

jukkar reviewed (commented):
Looks good, just minor comment about the doxygen comments

fengming-ye committed:
net: wifi: shell: add DPP commands

DPP shell command handler will parse user args to params in enum
and send l2 mgmt DPP requests.

Signed-off-by: Fengming Ye <frank.ye@nxp.com>
